qp_shotgun/kneaddata/kneaddata.py

```
@@ -0,0 +1,278 @@
+    parameters : dict
+        The command's parameters, keyed by parameter name
+
+    Returns
+    -------
+    cmds: list of str
+        The KneadData commands
+    samples: list of tup
+        list of 4-tuples with run prefix, sample name, fwd read fp, rev read fp
+
+    Notes
+    -----
+    Currently this is requiring matched pairs in the make_read_pairs_per_sample
+    step but implicitly allowing empty reverse reads in the actual command
+    generation. This behavior may allow support of situations with empty
+    reverse reads in some samples, for example after trimming and QC.
+    """
+    # we match filenames, samples, and run prefixes
+    samples = make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file)
+
+    cmds = []
+
+    param_string = _format_kneaddata_params(parameters)
+    for run_prefix, sample, f_fp, r_fp in samples:
+        r_fp_str = ' --input "%s"' % r_fp if r_fp is not None else ""
+        cmds.append('kneaddata --input "%s"%s --output "%s" '
+                    '--output-prefix "%s" %s' % (
+                        f_fp, r_fp_str, join(out_dir, run_prefix), run_prefix,
+                        param_string))
+
+    return cmds, samples
+
+
+def _run_commands(qclient, job_id, commands, msg):
+    for i, cmd in enumerate(commands):
+        qclient.update_job_step(job_id, msg % i)
+        std_out, std_err, return_value = system_call(cmd)
+        if return_value != 0:
+            error_msg = ("Error running KneadData:\nStd out: %s\nStd err: %s"
+                         "\n\nCommand run was:\n%s"
+                         % (std_out, std_err, cmd))
+            return False, error_msg
+
+    return True, ""
+
+
+def _per_sample_ainfo(out_dir, samples):
+    pf = []
+    r1f = []
+    r2f = []
+    for rp, _, _, _ in samples:
+        smd = partial(join, out_dir, rp)
+
+        fname = smd('%s.trimmed.1.fastq' % rp)
+        if not exists(fname):
+            raise ValueError("%s doesn't exist" % fname)
```

Do you want to raise an error if each of these doesn't exist? 

Maybe suggest adding a warning instead and creating an empty fastq, for example in a high-quality run where there are no unpaired forward reads recovered.

---

That makes sense, what about this: don't add to the artifact if it doesn't exist?

---

Sounds good to me, as long as downstream can handle elegantly. For example, in line 55 you check paired end reads by comparing length of lists -- wouldn't be able to take that approach after this step.

---

Good point, should we remove that check?

---

Ehhhh not here because I think this should be getting raw sequence, so definitely everything should be paired. But e.g. for Humann2 processing or the like.

Probably stuff that will be smoothed over by a more useful fastq representation / file format / object...

---

OK, let's do this. We will just add if present and after we check the full pipeline with real data we can make a decision, sounds good?

---

Sounds good to me!